src/pages/api/authOptions.ts

```
@@ -13,60 +13,101 @@ import { Adapter } from "next-auth/adapters";
+});
+
+const githubEnvParsed = githubEnvSchema.safeParse(process.env);
+const googleEnvParsed = googleEnvSchema.safeParse(process.env);
+const discordEnvParsed = discordEnvSchema.safeParse(process.env);
+
+const providers: Provider[] = [
+  CredentialsProvider({
+    credentials: {
+      // TODO: Change name to a unique identifier, maybe for email or both idk.
+      name: {
+        label: "Username",
+        type: "username",
+        placeholder: "Username",
+      },
+      password: {
+        label: "Password",
+        type: "password",
+        placeholder: "Password",
       },
-      async authorize(credentials) {
-        if (!credentials || !credentials.name || !credentials.password)
-          return null;
-
-        const creds = await loginSchema.parseAsync({
-          username: credentials.name,
-          password: credentials.password,
-        });
-
-        const dbUser = await prisma.user.findFirst({
-          where: { name: creds.username },
-        });
-
-        if (
-          dbUser &&
-          dbUser.password &&
-          (await compare(credentials.password, dbUser.password))
-        ) {
-          const dbUserWithoutPassword = {
-            name: dbUser.name,
-            email: dbUser.email,
-          };
-          return dbUserWithoutPassword as User;
-        }
+    },
+    async authorize(credentials) {
+      if (!credentials || !credentials.name || !credentials.password) {
+        return null;
+      }
+
+      const creds = await loginSchema.parseAsync({
+        username: credentials.name,
+        password: credentials.password,
+      });
+
+      const dbUser = await prisma.user.findFirst({
+        where: { name: creds.username },
+      });
 
+      if (!dbUser || !dbUser.password) {
         return null;
-      },
-    }),
+      }
+
+      const doPasswordsMatch = await compare(
+        credentials.password,
+        dbUser.password
+      );
+
+      if (doPasswordsMatch) {
```

i think code makes more sense with early-returns in "failure cases", related to the "fail fast" principle.
in practise, that'd mean `if (!doPasswordsMatch) { return null; } ...`